cache/compressed_secondary_cache.cc

```diff
@@ -145,6 +142,88 @@ std::string CompressedSecondaryCache::GetPrintableOptions() const {
   return ret;
 }
 
+std::unique_ptr<CompressedSecondaryCache::CacheValueChunk>
+CompressedSecondaryCache::SplitValueIntoChunks(
+    const Slice& value, const CompressionType compression_type,
+    size_t& charge) {
+  assert(!value.empty());
+  // If charge > the max size
+  const char* src_ptr = value.data();
+  size_t src_size{charge};
+  CacheAllocationPtr ptr;
+
+  std::unique_ptr<CacheValueChunk> head = std::make_unique<CacheValueChunk>();
+  CacheValueChunk* current_chunk = head.get();
+  // Do not split when value size is large or there is no compression.
+  if (src_size >= malloc_bin_sizes_.back() ||
+      compression_type == kNoCompression) {
+    charge += sizeof(*current_chunk);
+    ptr = AllocateBlock(src_size, cache_options_.memory_allocator.get());
+    memcpy(ptr.get(), src_ptr, src_size);
+    current_chunk->chunk_ptr = std::move(ptr);
+    current_chunk->charge = src_size;
+
+  } else {
+    while (src_size > 0) {
+      charge += sizeof(*current_chunk);
+      if (src_size <= malloc_bin_sizes_.front()) {
+        ptr = AllocateBlock(src_size, cache_options_.memory_allocator.get());
+        memcpy(ptr.get(), src_ptr, src_size);
+        current_chunk->chunk_ptr = std::move(ptr);
+        current_chunk->charge = src_size;
+        src_size = 0;
+      } else {
+        auto upper = std::upper_bound(malloc_bin_sizes_.begin(),
+                                      malloc_bin_sizes_.end(), src_size);
+        current_chunk->charge = *(--upper);
```

Comment:
If ```src_size``` is close to ```*upper```, i.e ```*upper - src_size < malloc_bin_sizes_.front()```, then allocating a chunk from ```upper``` bin would be optimal rather than splitting.

Reply:
Will address this in the next commit.